### PR TITLE
Update Synapse with ADF swagger changes/2019-06-01-preview

### PR DESCRIPTION
**Pull request by drosales57**
This is a PR generated at [OpenAPI Hub](https://aka.ms/openapihub). You can view your [work branch via this link](https://portal.azure-devex-tools.com/branch/drosales57/azure-rest-api-specs/darosale%2FrefreshAdfSwaggerForSynapse...master/).

### Changelog
Please ensure to add changelog with this PR by answering the following questions.
  1. What's the purpose of the update?    
      - [ ] new service onboarding 
      - [ ] new API version 
      - [ ] update existing version for new feature 
      - [ ] update existing version to fix swagger quality issue in s360
      - [ ] Other, please clarify 
  2. When you are targeting to deploy new service/feature to public regions? Please provide date, or month to public if date is not available yet.
  3. When you expect to publish swagger? Please provide date, or month to public if date is not available yet.
  4. If it's an update to existing version,  please select SDKs of specific language and CLIs that require refresh after swagger is published.
      - [ ] SDK of .NET (need service team to ensure code readiness)
      - [ ] SDK of Python
      - [ ] SDK of Java
      - [ ] SDK of Js
      - [ ] SDK of Go
      - [ ] PowerShell
      - [ ] CLI
      - [ ] Terraform
      - [ ] No, no need to refresh for updates in this PR

### Contribution checklist:
- [ ] I commit to follow the [Breaking Change Policy](http://aka.ms/bcforapi) of “no breaking changes
- [ ] I have reviewed the [documentation](https://aka.ms/ameonboard) for the workflow.
- [ ] [Validation tools](https://aka.ms/swaggertools) were run on swagger spec(s) and errors have all been fixed in this PR. [How to fix?](https://aka.ms/ci-fix)

If any further question about AME onboarding or validation tools, please view the [FAQ](https://aka.ms/faqinprreview).

### ARM API Review Checklist
- [ ] Ensure to check this box if one of the following scenarios meet updates in the PR, so that label “**WaitForARMFeedback**” will be added automatically to involve ARM API Review. Failure to comply may result in delays for manifest application. Note this does not apply to data plane APIs, all “removals” and “adding a new property” no more require ARM API review.
  - Adding new API(s)
  - Adding a new API version
  - Adding a new service

- [ ] Please ensure you've reviewed following [guidelines](https://aka.ms/rpguidelines) including [ARM resource provider contract](https://github.com/Azure/azure-resource-manager-rpc) and [REST guidelines](https://github.com/microsoft/api-guidelines/blob/vNext/azure/Guidelines.md). Estimated time (4 hours). This is required before you can request review from ARM API Review board.

- [ ] If you are blocked on ARM review and want to get the PR merged with urgency, please get the ARM oncall for reviews (*RP Manifest Approvers* team under <ins>Azure Resource Manager service</ins>) from IcM and reach out to them. 

### Breaking Change Review Checklist 
If there are following updates in the PR, ensure to request an approval from API Review Board as defined in the [Breaking Change Policy](http://aka.ms/bcforapi). 

- [ ] Removing API(s) in stable version
- [ ] Removing properties in stable version
- [ ] Removing API version(s) in stable version
- [ ] Updating API in stable version with Breaking Change Validation errors
- [ ] Updating API(s) in preview over 1 year

**Action**: to initiate an evaluation of the breaking change, create a new intake using the [template for breaking changes](https://msazure.visualstudio.com/One/_workitems/create/Scenario?templateId=d1c215ec-5d6a-4cb3-a193-95ddd0a1ba96&ownerId=85bb92fb-21fe-41ca-9309-47f2c5f57515). Addition details on the process and office hours are on the [Breaking change Wiki](https://dev.azure.com/msazure/AzureWiki/_wiki/wikis/AzureWiki.wiki/37684/Breaking-Changes).

Please follow the link to find more details on [PR review process](https://aka.ms/SwaggerPRReview).

### CODE REVIEW

**Review by ajay-arora — APPROVED** [review]

**Review by idear1203 — APPROVED** [review]
:shipit:

File: specification/synapse/data-plane/Microsoft.Synapse/preview/2019-06-01-preview/entityTypes/Trigger.json [hunk 1]
```diff
@@ -335,12 +335,55 @@
           "Microsoft.Storage.BlobDeleted"
         ],
         "x-ms-enum": {
-          "name": "BlobEventType",
+          "name": "BlobEventTypes",
```
Comment (idear1203) [inline specification/synapse/data-plane/Microsoft.Synapse/preview/2019-06-01-preview/entityTypes/Trigger.json:338]: >BlobEventTypes [](start = 19, length = 14)

Can we keep the previous `BlobEventType` ? I made that change before to fix the CI pipeline issue:

```
"readme":"synapse/data-plane/readme.md",
"tag":"package-artifacts-2019-06-01-preview",
"details":"Duplicate Schema named BlobEventTypes -- type: undefined => "string",
items: "$ref":"#/components/schemas/BlobEventTypes" =>,
enum: undefined => ["Microsoft.Storage.BlobCreated",
"Microsoft.Storage.BlobDeleted"],
x-ms-enum: undefined => "name":"BlobEventTypes",
"modelAsString":true ; This error can be temporarily avoided by using the 'modelerfour.lenient-model-deduplication' setting. NOTE: This setting will be removed in a future version of @autorest/modelerfour; schemas should be updated to fix this issue sooner than that."
```

Comment (drosales57) [reply specification/synapse/data-plane/Microsoft.Synapse/preview/2019-06-01-preview/entityTypes/Trigger.json]: done

File: specification/synapse/data-plane/Microsoft.Synapse/preview/2019-06-01-preview/entityTypes/Dataset.json [hunk 2]
```diff
@@ -612,17 +753,19 @@
           "description": "The code page name of the preferred encoding. If miss, the default value is UTF-8, unless BOM denotes another Unicode encoding. Refer to the name column of the table in the following link to set supported values: https://msdn.microsoft.com/library/system.text.encoding.aspx. Type: string (or Expression with resultType string)."
         },
         "compressionCodec": {
-          "type": "string",
+          "type": "object",
```
Comment (joheredi) [inline specification/synapse/data-plane/Microsoft.Synapse/preview/2019-06-01-preview/entityTypes/Dataset.json:756]: What is the reason for this string changing to an object?

File: specification/synapse/data-plane/Microsoft.Synapse/preview/2019-06-01-preview/entityTypes/Dataset.json [hunk 3]
```diff
@@ -552,21 +691,22 @@
...
           "enum": [
             "none",
             "gzip",
             "snappy",
             "lzo"
           ],
           "x-ms-enum": {
-            "name": "ParquetCompressionCodec",
+            "name": "compressionCodec",
```
Comment (joheredi) [inline specification/synapse/data-plane/Microsoft.Synapse/preview/2019-06-01-preview/entityTypes/Dataset.json:709]: I see a few enums with the same names but different values. It feels like if they are different they should have different names. Why is it changing from `ParquetCompressionCodec` to `compressionCodec`

File: specification/synapse/data-plane/Microsoft.Synapse/preview/2019-06-01-preview/entityTypes/Dataset.json [hunk 4]
```diff
@@ -688,6 +832,48 @@
...
+    "XmlDatasetTypeProperties": {
+      "description": "Xml dataset properties.",
+      "type": "object",
+      "properties": {
+        "location": {
+          "$ref": "#/definitions/DatasetLocation",
+          "description": "The location of the json data storage."
+        },
+        "encodingName": {
+          "type": "object",
```
Comment (joheredi) [inline specification/synapse/data-plane/Microsoft.Synapse/preview/2019-06-01-preview/entityTypes/Dataset.json:861]: Is this really an object? It seems like encodingName should be a string

File: specification/synapse/data-plane/Microsoft.Synapse/preview/2019-06-01-preview/entityTypes/Dataset.json [hunk 5]
```diff
@@ -688,6 +832,48 @@
...
+          "description": "The code page name of the preferred encoding. If not specified, the default value is UTF-8, unless BOM denotes another Unicode encoding. Refer to the name column of the table in the following link to set supported values: https://msdn.microsoft.com/library/system.text.encoding.aspx. Type: string (or Expression with resultType string)."
+        },
+        "nullValue": {
+          "type": "object",
+          "description": "The null value string. Type: string (or Expression with resultType string)."
```
Comment (joheredi) [inline specification/synapse/data-plane/Microsoft.Synapse/preview/2019-06-01-preview/entityTypes/Dataset.json:866]: same here, this seems like it should be a string

File: specification/synapse/data-plane/Microsoft.Synapse/preview/2019-06-01-preview/entityTypes/LinkedService.json [hunk 6]
```diff
@@ -532,23 +558,23 @@
             "Online",
             "OnPremisesWithIfd"
           ],
-          "type": "string",
```
Comment (joheredi) [inline specification/synapse/data-plane/Microsoft.Synapse/preview/2019-06-01-preview/entityTypes/LinkedService.json:535]: What's the reason of this change?

File: specification/synapse/data-plane/Microsoft.Synapse/preview/2019-06-01-preview/entityTypes/Trigger.json [hunk 7]
```diff
@@ -576,17 +618,18 @@
               "format": "date-time",
               "description": "The end time for the time period for which restatement is initiated. Only UTC time is currently supported."
             },
-            "maxConcurrency": {
+            "rerunConcurrency": {
```
Comment (joheredi) [inline specification/synapse/data-plane/Microsoft.Synapse/preview/2019-06-01-preview/entityTypes/Trigger.json:621]: I'm curious about this rename. Why did you drop `maxConcurrency`? IMO the previous name represents better the semantics of this property. Also having rerun in this properties name feels redundant with the parent name

File: specification/synapse/data-plane/Microsoft.Synapse/preview/2019-06-01-preview/entityTypes/Dataset.json (hunk 2 above)
Comment (drosales57) [reply specification/synapse/data-plane/Microsoft.Synapse/preview/2019-06-01-preview/entityTypes/Dataset.json:756]: Hi, we are syncing with the Azure Data Factory(ADF) swagger. The Synapse-side swagger is out of date with Azure Data Factory counterpart. This has led to a lot feature gaps Synapse customers.

Comment (drosales57) [reply specification/synapse/data-plane/Microsoft.Synapse/preview/2019-06-01-preview/entityTypes/Dataset.json:756]: Here is the link to the ADF public swagger 
https://github.com/Azure/azure-rest-api-specs/blob/master/specification/datafactory/resource-manager/Microsoft.DataFactory/stable/2018-06-01/entityTypes/Dataset.json

File: specification/synapse/data-plane/Microsoft.Synapse/preview/2019-06-01-preview/entityTypes/Dataset.json (hunk 3 above)
Comment (drosales57) [reply specification/synapse/data-plane/Microsoft.Synapse/preview/2019-06-01-preview/entityTypes/Dataset.json:709]: This was renamed in ADF swagger so we are just trying to keep both Synapse and ADF in sync

File: specification/synapse/data-plane/Microsoft.Synapse/preview/2019-06-01-preview/entityTypes/Dataset.json [hunk 8]
```diff
@@ -472,10 +472,36 @@
...
+        "level": {
+          "description": "The TarGZip compression level.",
+          "$ref": "#/definitions/CompressionLevel"
+        }
+      }
+    },
     "CompressionLevel": {
       "title": "Microsoft.Azure.Management.DataFactories.Models.CompressionLevel",
       "description": "All available compression levels.",
-      "type": "string",
```
Comment (lmazuel) [inline specification/synapse/data-plane/Microsoft.Synapse/preview/2019-06-01-preview/entityTypes/Dataset.json:478]: This should stay a string

File: specification/synapse/data-plane/Microsoft.Synapse/preview/2019-06-01-preview/entityTypes/Dataset.json [hunk 9]
```diff
@@ -552,21 +691,22 @@
...
       "description": "Parquet dataset properties.",
+      "type": "object",
       "properties": {
         "location": {
           "$ref": "#/definitions/DatasetLocation",
           "description": "The location of the parquet storage."
         },
         "compressionCodec": {
-          "type": "string",
+          "type": "object",
```
Comment (lmazuel) [inline specification/synapse/data-plane/Microsoft.Synapse/preview/2019-06-01-preview/entityTypes/Dataset.json:701]: This is not an object, this is a string

File: specification/synapse/data-plane/Microsoft.Synapse/preview/2019-06-01-preview/entityTypes/Dataset.json (hunk 2 above)
Comment (lmazuel) [reply specification/synapse/data-plane/Microsoft.Synapse/preview/2019-06-01-preview/entityTypes/Dataset.json:756]: This was an incorrect change in ADF in the first place, this is not an object. Sorry, but we should not just accept it here because it's copy/pasting :(. We shoudl fix ADF Swagger instead.

File: specification/synapse/data-plane/Microsoft.Synapse/preview/2019-06-01-preview/entityTypes/Dataset.json [hunk 10]
```diff
@@ -1078,13 +1326,92 @@
...
+          "$ref": "#/definitions/AzureDataLakeStoreDatasetTypeProperties"
+        }
+      }
+    },
+    "AzureDataLakeStoreDatasetTypeProperties": {
+      "description": "Azure Data Lake Store dataset properties.",
+      "type": "object",
+      "properties": {
+        "folderPath": {
+          "type": "object",
```
Comment (lmazuel) [inline specification/synapse/data-plane/Microsoft.Synapse/preview/2019-06-01-preview/entityTypes/Dataset.json:1359]: That seems inconsistent with the description that says it's a string

File: specification/synapse/data-plane/Microsoft.Synapse/preview/2019-06-01-preview/entityTypes/Dataset.json (hunk 2 above)
Comment (idear1203) [reply specification/synapse/data-plane/Microsoft.Synapse/preview/2019-06-01-preview/entityTypes/Dataset.json:756]: Agree. For enumeration, the type should be string. During code generation, I can see some warnings if type if object. @drosales57 , could you please at least fix this on Synapse side?

File: specification/synapse/data-plane/Microsoft.Synapse/preview/2019-06-01-preview/entityTypes/Dataset.json (hunk 3 above)
Comment (idear1203) [reply specification/synapse/data-plane/Microsoft.Synapse/preview/2019-06-01-preview/entityTypes/Dataset.json:709]: I suspect this is bug in ADF Swagger spec because I can still see "avroCompressionCodec", "orcCompressionCodec" is the Swagger spec. It doesn't make much sense to me that we only change the "parquet" one. I guess in the original ADF change, the developer wanted to change it to be "parquetCompressionCodec". @drosales57 , could you please check this with someone in ADF team?

File: specification/synapse/data-plane/Microsoft.Synapse/preview/2019-06-01-preview/entityTypes/Dataset.json (hunk 4 above)
Comment (idear1203) [reply specification/synapse/data-plane/Microsoft.Synapse/preview/2019-06-01-preview/entityTypes/Dataset.json:861]: As far as I know, some properties on ADF side can accept both strings and objects. I am not sure this is such case or not. Seems not. @drosales57 , can we change it to string type?

File: specification/synapse/data-plane/Microsoft.Synapse/preview/2019-06-01-preview/entityTypes/LinkedService.json (hunk 6 above)
Comment (idear1203) [reply specification/synapse/data-plane/Microsoft.Synapse/preview/2019-06-01-preview/entityTypes/LinkedService.json:535]: Agree. For enumerations, we should use string.

File: specification/synapse/data-plane/Microsoft.Synapse/preview/2019-06-01-preview/entityTypes/Trigger.json (hunk 7 above)
Comment (idear1203) [reply specification/synapse/data-plane/Microsoft.Synapse/preview/2019-06-01-preview/entityTypes/Trigger.json:621]: @drosales57 might be able to provide some explanations. But can we keep this change anyway for consistency reason? Would it be a good idea to use "x-ms-client-name" to rename it to "maxConcurrency" at SDK side?

File: specification/synapse/data-plane/Microsoft.Synapse/preview/2019-06-01-preview/entityTypes/Dataset.json (hunk 10 above)
Comment (idear1203) [reply specification/synapse/data-plane/Microsoft.Synapse/preview/2019-06-01-preview/entityTypes/Dataset.json:1359]: @drosales57 I also think string would be a better choice here. Is it possible to use string?

**Comment by drosales57** [comment]
@lmazuel @joheredi Please see my response to your comment about using string type vs object type. In summary, this is intended for ADF expression support

**Review by idear1203 — APPROVED** [review]
Talked to @drosales57 , we would like to change the enums to string types but leave non-enums to be object types. Those non-enums can be other expressions or strings.

File: specification/synapse/data-plane/Microsoft.Synapse/preview/2019-06-01-preview/entityTypes/LinkedService.json [hunk 11]
```diff
@@ -561,7 +587,7 @@
             "Ifd",
             "AADServicePrincipal"
           ],
-          "type": "string",
+          "type": "object",
```
Comment (idear1203) [inline specification/synapse/data-plane/Microsoft.Synapse/preview/2019-06-01-preview/entityTypes/LinkedService.json:590]: >type": "object", [](start = 11, length = 16)

string?

File: specification/synapse/data-plane/Microsoft.Synapse/preview/2019-06-01-preview/entityTypes/LinkedService.json [hunk 12]
```diff
@@ -624,6 +650,7 @@
     },
     "DynamicsCrmLinkedServiceTypeProperties": {
       "description": "Dynamics CRM linked service properties.",
+      "type": "object",
```
Comment (idear1203) [inline specification/synapse/data-plane/Microsoft.Synapse/preview/2019-06-01-preview/entityTypes/LinkedService.json:653]: >"type": "object", [](start = 6, length = 17)

string

File: specification/synapse/data-plane/Microsoft.Synapse/preview/2019-06-01-preview/entityTypes/Dataset.json [hunk 13]
```diff
@@ -485,6 +512,67 @@
...
+          "description": "The version for the S3 object. Type: string (or Expression with resultType string)."
+        },
+        "modifiedDatetimeStart": {
+          "type": "object",
+          "description": "The start of S3 object's modified datetime. Type: string (or Expression with resultType string)."
+        },
+        "modifiedDatetimeEnd": {
+          "type": "object",
+          "description": "The end of S3 object's modified datetime. Type: string (or Expression with resultType string)."
+        },
```
Comment (joheredi) [inline specification/synapse/data-plane/Microsoft.Synapse/preview/2019-06-01-preview/entityTypes/Dataset.json:562]: All of these should be `string` and not `object`

File: specification/synapse/data-plane/Microsoft.Synapse/preview/2019-06-01-preview/entityTypes/Dataset.json [hunk 14]
```diff
@@ -533,6 +622,57 @@
...
+          "type": "object",
+          "description": "The sheet of excel file. Type: string (or Expression with resultType string)."
+        },
+        "range": {
+          "type": "object",
+          "description": "The partial data of one sheet. Type: string (or Expression with resultType string)."
+        },
+        "firstRowAsHeader": {
+          "type": "object",
+          "description": "When used as input, treat the first row of data as headers. When used as output,write the headers into the output as the first row of data. The default value is false. Type: boolean (or Expression with resultType boolean)."
```
Comment (joheredi) [inline specification/synapse/data-plane/Microsoft.Synapse/preview/2019-06-01-preview/entityTypes/Dataset.json:660]: All of these should be `string` and not `object`. firstRowHeader should be `boolean`

File: specification/synapse/data-plane/Microsoft.Synapse/preview/2019-06-01-preview/entityTypes/Dataset.json [hunk 15]
```diff
@@ -533,6 +622,57 @@
...
+        },
+        "compression": {
+          "description": "The data compression method used for the json dataset.",
+          "$ref": "#/definitions/DatasetCompression"
+        },
+        "nullValue": {
+          "type": "object",
```
Comment (joheredi) [inline specification/synapse/data-plane/Microsoft.Synapse/preview/2019-06-01-preview/entityTypes/Dataset.json:667]: This should be string

File: specification/synapse/data-plane/Microsoft.Synapse/preview/2019-06-01-preview/entityTypes/Dataset.json [hunk 16]
```diff
@@ -763,6 +952,57 @@
...
+          "description": "The name of the Azure Blob. Type: string (or Expression with resultType string)."
+        },
+        "modifiedDatetimeStart": {
+          "type": "object",
+          "description": "The start of Azure Blob's modified datetime. Type: string (or Expression with resultType string)."
+        },
+        "modifiedDatetimeEnd": {
+          "type": "object",
+          "description": "The end of Azure Blob's modified datetime. Type: string (or Expression with resultType string)."
+        },
```
Comment (joheredi) [inline specification/synapse/data-plane/Microsoft.Synapse/preview/2019-06-01-preview/entityTypes/Dataset.json:995]: All of these should be `string`

**Comment by joheredi** [comment]
> Talked to @drosales57 , we would like to change the enums to string types but leave non-enums to be object types. Those non-enums can be other expressions or strings.

@drosales57, @idear1203 I would like to understand better the scenario when `objects` are used instead of `string`, would you mind sharing an example of a request/response with an `object` vs `string`? Is there a more specific object definition we can use instead of just `object`?

**Comment by drosales57** [comment]
> > Talked to @drosales57 , we would like to change the enums to string types but leave non-enums to be object types. Those non-enums can be other expressions or strings.
> 
> @drosales57, @idear1203 I would like to understand better the scenario when `objects` are used instead of `string`, would you mind sharing an example of a request/response with an `object` vs `string`? Is there a more specific object definition we can use instead of just `object`?


Many fields in datasets, linkedservices, and pipeline activities can have an expresssion or static value like an integer or stirng.
Let take a simple example of a pipeline with a Wait activity. The wait activity waits for a specified number seconds (has type int)

Here's the sample payload that we would publish
```
{
    "name": "Pipeline1",
    "properties": {
      "activities": [
        {
          "name": "Wait1",
          "type": "Wait",
          "dependsOn": [],
          "userProperties": [],
          "typeProperties": { "waitTimeInSeconds": 1 }
        }
      ],
      "annotations": []
    }
  }
```


Now say the customer wants to vary the wait time in the Wait activity based on a pipeline parameter value. Below we have pipeline parameter called "waitTimeParameter"
```
{
    "name": "Pipeline1",
    "properties": {
      "activities": [
        {
          "name": "Wait1",
          "type": "Wait",
          "dependsOn": [],
          "userProperties": [],
          "typeProperties": {
            "waitTimeInSeconds": {
              "value": "@pipeline().parameters.waitTimeParameter",
              "type": "Expression"
            }
          }
        }
      ],
      "parameters": { "waitTimeParameter": { "type": "int" } },
      "annotations": []
    }
  }
```


We can reference this pipeline parameter value using the expression (type object):
```
{
    "value": "@pipeline().parameters.waitTimeParameter",
    "type": "Expression"
}
```

**Review by lmazuel — APPROVED** [review]
